### ISSUE TITLE
Very slow trimmed surface export

### ISSUE THREAD

**Issue by jeffallen** [OpenVSP/OpenVSP#272]
In the attached [issue272.zip](https://github.com/OpenVSP/OpenVSP/files/12371391/issue272.zip), there is a set called "export". It has two elements in it.

1. Use the Set Editor to uncheck the tail boom
2. use Analysis -> Trimmed Surfaces, set "use set" to "export", and "degen use set" to None.
3. Click "intersect and export". The fuselage is exported correctly.
4. In the set editor, enable "tail boom" in set "export".
5. Now clicking on "Intersect and Export" will take much longer, on the order of 30 minutes.

The long step is "Binary Adaptation Curve Approximation".

Also: Where is the difference between "Use Set" and "Degen Use Set" explained?

**Comment by ramcdona** [comment]
On my computer, while it paused for a moment at that point, it eventually continued and eventually completed the intersect and export.

There is a known performance problem in BuildChains that I will hopefully have some time to look into soon.

I believe this performance problem is made worse by the subsurfaces on your fuselage (though it really shouldn't be this bad).  If you don't need them in your exported STEP file, you might try turning them off (disable the subsurfaces on the first tab of the Intersect And Export GUI).

All the components in the Normal Set will be included as normal thick-surface representations of the shape.

All components in the Degen Set will be included as a thin-surface representation -- a mean camber line for wings and a cruciform for Fuselages.  You probably don't want this.  So set the Degen Use Set to None and you'll be fine.

The tail boom geometry is open at the front and the aft -- this has the potential of confusing the inside/outside tests and might cause problems down the road.  I recommend you keep every component individually watertight.

Also, this geometry could probably effectively be modeled with a single geometry -- where the boom is modeled as an extension of the fuselage -- all in one component, no intersection required.

Finally, I made a fix in another branch that might be useful to you (I'm not sure).  https://github.com/ramcdona/OpenVSP/commit/ebf605071217f309b6a6d47f5a22514504935da9

**Comment by jeffallen** [comment]
I agree that if you leave it a long time, it does finish. Sorry to get impatient. (I updated the issue with this.)

I am using this as a case to learn how use Windows Performance Recorder, so that I can post profiling info here.

As for removing subsurfaces, etc, yes... but no. It already took me a certain amount of debugging to even find out why these two volumes cause the export to grind to a halt. We have many other volumes in our model, and it took a process of elimination to find this problem. I'd like to work hard on solving this one, in case the same problem is hitting us on other parts of our model and we just haven't noticed yet.

I am now looking into the watertightness issue.

**Comment by jeffallen** [comment]
I have tried turning off "Intersect Subsurfaces" and it didn't make a difference.

I have not been able to get any profiles with any smoking guns. I did see a significant amount of time in new/del/realloc/malloc, etc. I suspect that there's an opportunity someplace in there to pre-calculate some objects and then reuse them instead of re-creating them. The profile I was looking at took me to IntersectPatch.cpp, line 67, where there are lots of SurfPatches that get created, used and deleted. To be clear, I don't have any actionable suggestions yet, just a spidey-sense.

**Comment by ramcdona** [comment]
If you go in and un-comment lines 53-55 here https://github.com/OpenVSP/OpenVSP/blob/main/src/cfd_mesh/SurfaceIntersectionMgr.h#L53

It will turn on some more verbose messages including simple timing of every step of the process.

I really don't see why your case is so slow.  There really isn't anything that special or complex about it.

I tried approximating your super ellipses with rounded rectangles.  And then getting rid of setting curvature in skinning.  And also turning off the subsurface intersection -- none of that made much difference...

I am scheduled to spend some time working on the performance of these algorithms soon, so it will be good to have some test cases.

**Comment by jeffallen** [comment]
Here is MeshDebug/[log.txt](https://github.com/OpenVSP/OpenVSP/files/12408948/log.txt).

I unfortunately lost the timing info that was printed in the UI, but it was nothing special... ms for the first few jobs, and then lots and lots of seconds for the final ones starting with BuildChains.

I also needed this code change to get it to compile:
```
diff --git a/src/cfd_mesh/SurfaceIntersectionMgr.cpp b/src/cfd_mesh/SurfaceIntersectionMgr.cpp
index fe524f7f..2b19f99a 100644
--- a/src/cfd_mesh/SurfaceIntersectionMgr.cpp
+++ b/src/cfd_mesh/SurfaceIntersectionMgr.cpp
@@ -18,7 +18,7 @@
 #include "MeshAnalysis.h"

 #ifdef DEBUG_CFD_MESH
-// #include <direct.h>
+#include <chrono>
 #include <sys/stat.h>
 #include <sys/types.h>
 #endif
```

**Comment by ramcdona** [comment]
I've made a lot of changes in this area -- driven by profiling.  It should be considerably faster.  I just ran your test case (attached to this issue) and it seemed to run in a reasonable amount of time.

These fixes will go into 3.37.0 -- which will be released in the next few days.

I'm going to close this issue (assuming it fixes things for you).  If you are still having problems, either re-open this issue, or open a new one.
